pkg/ccl/backupccl/backup_processor.go

```diff
@@ -312,23 +317,71 @@ func runBackupProcessor(
 	clusterSettings := flowCtx.Cfg.Settings
 
 	totalSpans := len(spec.Spans) + len(spec.IntroducedSpans)
-	todo := make(chan spanAndTime, totalSpans)
-	var spanIdx int
-	for _, s := range spec.IntroducedSpans {
-		todo <- spanAndTime{
-			spanIdx: spanIdx, span: s, firstKeyTS: hlc.Timestamp{}, start: hlc.Timestamp{},
-			end: spec.BackupStartTime,
+	requestSpans := make([]spanAndTime, 0, totalSpans)
+	rangeSizedSpans := preSplitExports.Get(&flowCtx.EvalCtx.Settings.SV)
+
+	for _, fullSpan := range spec.IntroducedSpans {
+		remainingSpan := fullSpan
+
+		if rangeSizedSpans {
+			rdi, err := flowCtx.Cfg.ExecutorConfig.(*sql.ExecutorConfig).RangeDescIteratorFactory.NewIterator(ctx, fullSpan)
+			if err != nil {
+				return err
+			}
+			for ; rdi.Valid(); rdi.Next() {
+				rangeDesc := rdi.CurRangeDescriptor()
+				subspan := remainingSpan.Intersect(roachpb.Span{Key: rangeDesc.StartKey.AsRawKey(), EndKey: rangeDesc.EndKey.AsRawKey()})
+
+				requestSpans = append(requestSpans, spanAndTime{
+					span: subspan, firstKeyTS: hlc.Timestamp{}, start: hlc.Timestamp{},
+					end: spec.BackupStartTime,
+				})
+				remainingSpan.Key = subspan.EndKey
+			}
+
+			if !remainingSpan.Overlaps(fullSpan) {
+				continue
+			}
 		}
-		spanIdx++
+
+		requestSpans = append(requestSpans, spanAndTime{
+			span: remainingSpan, firstKeyTS: hlc.Timestamp{}, start: hlc.Timestamp{},
+			end: spec.BackupStartTime,
+		})
 	}
-	for _, s := range spec.Spans {
-		todo <- spanAndTime{
-			spanIdx: spanIdx, span: s, firstKeyTS: hlc.Timestamp{}, start: spec.BackupStartTime,
-			end: spec.BackupEndTime,
+
+	for _, fullSpan := range spec.Spans {
+		remainingSpan := fullSpan
+
+		if rangeSizedSpans {
+			rdi, err := flowCtx.Cfg.ExecutorConfig.(*sql.ExecutorConfig).RangeDescIteratorFactory.NewIterator(ctx, fullSpan)
+			if err != nil {
+				return err
+			}
+			for ; rdi.Valid(); rdi.Next() {
+				rangeDesc := rdi.CurRangeDescriptor()
+				subspan := remainingSpan.Intersect(roachpb.Span{Key: rangeDesc.StartKey.AsRawKey(), EndKey: rangeDesc.EndKey.AsRawKey()})
```

Comment:
"It should never happen" but do we want to check that our subspan is valid here?

Reply:
made it an error